homeassistant/helpers/entity.py

```diff
@@ -642,23 +645,57 @@ def _async_write_ha_state(self) -> None:
         if DATA_CUSTOMIZE in self.hass.data:
             attr.update(self.hass.data[DATA_CUSTOMIZE].get(self.entity_id))
 
-        # Convert temperature if we detect one
-        try:
+        def _convert_temperature(state: str, attr: dict) -> str:
+            # Convert temperature if we detect one
+            # pylint: disable-next=import-outside-toplevel
+            from homeassistant.components.sensor import SensorEntity
+
             unit_of_measure = attr.get(ATTR_UNIT_OF_MEASUREMENT)
             units = self.hass.config.units
-            domain = split_entity_id(self.entity_id)[0]
-            if (
-                unit_of_measure in (TEMP_CELSIUS, TEMP_FAHRENHEIT)
-                and unit_of_measure != units.temperature_unit
-                and domain != "sensor"
+            if unit_of_measure == units.temperature_unit or unit_of_measure not in (
+                TEMP_CELSIUS,
+                TEMP_FAHRENHEIT,
             ):
+                return state
+
+            domain = split_entity_id(self.entity_id)[0]
+            if domain != "sensor":
+                if not self._temperature_reported:
+                    self._temperature_reported = True
+                    report_issue = self._suggest_report_issue()
+                    _LOGGER.warning(
+                        "Entity %s (%s) does not convert temperatures, this will "
+                        "be unsupported in Home Assistant Core 2022.7. Please %s",
```

Comment:
```suggestion
                        "Entity %s (%s) relies on automatic temperature conversion, this will "
                        "be unsupported in Home Assistant Core 2022.7. Please %s",
```